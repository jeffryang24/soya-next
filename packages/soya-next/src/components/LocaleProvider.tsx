import React from "react";
import LocaleContext from "../i18n/LocaleContext";
import type { SoyaNextLocale } from "..";

export interface LocaleProviderProps {
  children: React.ReactNode;
  defaultLocale?: string;
  siteLocales?: string[];
  locale?: SoyaNextLocale;
}
export interface LocaleProviderState {
  defaultLocale?: string;
  siteLocales?: string[];
  locale?: SoyaNextLocale;
}

class LocaleProvider extends React.Component<
  LocaleProviderProps,
  LocaleProviderState
> {
  constructor(props: LocaleProviderProps) {
    super(props);
    this.state = {
      defaultLocale: props.defaultLocale,
      siteLocales: props.siteLocales,
      locale: props.locale,
    };
  }

  static getDerivedStateFromProps(
    props: LocaleProviderProps,
    state: LocaleProviderState
  ) {
    if (props.locale !== state.locale) {
      return { locale: props.locale };
    }
    return null;
  }

  render() {
    return (
      <LocaleContext.Provider value={this.state}>
        {React.Children.only(this.props.children)}
      </LocaleContext.Provider>
    );
  }
}

export default LocaleProvider;
