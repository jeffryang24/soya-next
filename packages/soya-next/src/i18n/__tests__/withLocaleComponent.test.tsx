import React from "react";
import { mount } from "enzyme";
import withLocale from "../withLocaleComponent";
import LocaleProvider from "../../components/LocaleProvider";

describe("withLocale", () => {
  function setupTest() {
    const context = {
      locale: {
        language: "id",
        country: "id",
      },
      defaultLocale: "id-id",
      siteLocales: ["id-id", "en-id"],
    };
    const Component = () => null;
    const ComponentWithLocale = withLocale(Component);

    return { context, Component, ComponentWithLocale };
  }

  it("should add default locale, locale, and site locales to component props", () => {
    const { context, Component, ComponentWithLocale } = setupTest();
    const wrapper = mount(
      <LocaleProvider {...context}>
        <ComponentWithLocale />
      </LocaleProvider>
    );
    expect(wrapper.find(Component).props()).toMatchSnapshot();
  });

  it("should override locale context with locale props", () => {
    const { context, Component, ComponentWithLocale } = setupTest();
    const locale = { language: "en", country: "id" };
    const wrapper = mount(
      <LocaleProvider {...context}>
        <ComponentWithLocale locale={locale} />
      </LocaleProvider>
    );
    expect(wrapper.find(Component).props()).toMatchSnapshot();
  });

  it("should accept locale string props", () => {
    const { context, Component, ComponentWithLocale } = setupTest();
    const wrapper = mount(
      <LocaleProvider {...context}>
        <ComponentWithLocale locale="en-id" />
      </LocaleProvider>
    );
    expect(wrapper.find(Component).props()).toMatchSnapshot();
  });
});
