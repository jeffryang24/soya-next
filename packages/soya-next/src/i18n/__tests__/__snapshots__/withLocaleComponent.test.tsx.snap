// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withLocale should accept locale string props 1`] = `
Object {
  "defaultLocale": "id-id",
  "locale": Object {
    "country": "id",
    "language": "en",
  },
  "siteLocales": Array [
    "id-id",
    "en-id",
  ],
}
`;

exports[`withLocale should add default locale, locale, and site locales to component props 1`] = `
Object {
  "defaultLocale": "id-id",
  "locale": Object {
    "country": "id",
    "language": "id",
  },
  "siteLocales": Array [
    "id-id",
    "en-id",
  ],
}
`;

exports[`withLocale should override locale context with locale props 1`] = `
Object {
  "defaultLocale": "id-id",
  "locale": Object {
    "country": "id",
    "language": "en",
  },
  "siteLocales": Array [
    "id-id",
    "en-id",
  ],
}
`;
